src/Disciplina/WorldState/Internal.hs

```diff
@@ -0,0 +1,495 @@
+
+import qualified Data.Tree.AVL as AVL
+import qualified Debug.Trace   as Debug
+
+-- | Stub representation of entity.
+data Entity = Entity Int
+    deriving (Show, Eq, Ord, Bounded, Generic, Binary, Default)
+
+-- | Global state of the network, maintained by each server node.
+data WorldState
+  = WorldState
+    { _accounts       :: AVLMap (Account Hash)
+    , _publications   :: AVLMap Publication
+    , _specalizations :: AVLMap DAG
+    , _storage        :: AVLMap Storage
+    , _code           :: AVLMap Code
+    }
+
+type AVLMap v = AVL.Map Hash Entity v
+
+-- | Proof of any action on global data.
+data WorldStateProof
+    = WorldStateProof
+        { _accountsProof      :: Proof (Account Hash)
+        , _publicationsProof  :: Proof Publication
+        , _specalizatiosProof :: Proof DAG
+        , _storageProof       :: Proof Storage
+        , _codeProof          :: Proof Code
+        }
+    deriving (Show, Generic, Binary)
+
+instance Eq WorldStateProof where
+    WorldStateProof a b c d e == WorldStateProof f g h i k =
+        and [a ==? f, b ==? g, c ==? h, d ==? i, e ==? k]
+      where
+        (==?) :: (Show s, Eq s, Binary s) => Proof s -> Proof s -> Bool
+        (==?) = (==) `on` getHash
+
+        -- TODO: move into Data.Tree.AVL.Proof
+        getHash proof = proof
+            ^.AVL._Proof    -- get the tree
+             .AVL.rootHash  -- reduce to hash
+
+type Proof = AVL.Proof Hash Entity
+
+type DAG         = Hash
+type Publication = Hash
+type Storage     = Hash
+type Code        = Hash
+
+-- | The changes the transaction consists from.
+data Change
+    = TransferTokens Entity      Amount
+    | Publicate      Publication
+    | CreateAccount  Entity      Code
+    deriving (Show)
+
+-- TODO: implement signing (and make 'Entity' to be actual public key).
+data Transaction = Transaction
+    { _tAuthor     :: Entity
+    , _tChanges    :: [Change]
+    , _tNonce      :: Int
+    }
+
+instance Show Transaction where
+    show (Transaction who what nonce) =
+        concat
+            [ show who
+            , ": "
+            , show what
+            , " ("
+            , show nonce
+            , ")"
+            ]
+
+-- | Wrapper that adds proof to either block or transaction.
+data WithProof a = WithProof
+    { _wpBody    :: a
+    , _wpProof   :: WorldStateProof
+    , _wpEndHash :: Hash
+    }
+
+-- | World state on server.
+data Server = Server
+    { _sWorld :: WorldState
+    }
+
+-- | World state on client.
+data Client = Client
+    { _cProof :: WorldStateProof
+    }
+
+-- | Set of node identities to generate proof from. Proof prefab.
+data RevisionSets = RevisionSets
+    { _accountRevSet       :: AVL.RevSet
+    , _publicationRevSet   :: AVL.RevSet
+    , _specalizationRevSet :: AVL.RevSet
+    , _storageRevSet       :: AVL.RevSet
+    , _codeRevSet          :: AVL.RevSet
+    }
+    deriving (Default, Generic)
+
+data Environment = Environment
+    { _eAuthor :: Entity
+    }
```

Comment:
No, it is a _local_ configuration. It carries the identity of author of the changes. It can (and will) change on each transaction generated/consumed for a period of time, so the `Reader` is fine abstraction in my eyes.

Reply:
Ah, okay, so it seems to me that I pretty much missed the point of `WorldM` and stuff. I imagined that `runWorldM` is done on start and afterwards all the actions are performed inside the monad, and in your universe `runWorldM` is called every time for each separate action.

In this way, even `impersonate` seems kinda legit. Though it means that `WorldM` should be actually a `WorldT`, a transformer, which applies on top of some IO-based monad.